### PR TITLE
[Transform] improve performance by using point in time API for search

### LINKED ISSUE #73481: [Transform] Use point in time search, optimize query execution
Transform uses a composite aggregation to page through the source. Benchmarks have shown that the executed ordinary searches slow down the ingest of the source. Due to transform a `refresh` is triggered more often than usual, causing the source ingest to do more work than without it.

By using a point in time reader, transform will prevent causing so much churn (benchmarks have shown a potential reduction of refreshes by 50%).

### Requirements

 - open a _pit_ reader at the beginning of a new checkpoint
 - use _pit_ for all searches in the checkpoint
 	- re-create a _pit_ if necessary (e.g. timeout, start/stop)
 	- don't fail due to a broken _pit_ 
 	- `keepAlive` should be kept reasonably small
 - explicitly delete the _pit_ 
    - after a checkpoint
    - on stop
 - fallback to non _pit_ mode, in case a node older `7.10` is part of the local or remote (CCS) cluster

### Design considerations

Searches are executed by `ClientTransformIndexer` which inherits from `TransformIndexer`, adding the search capabilities given a `client`. This needs to be enhanced to create/destroy the _pit_ object.

Future: Given checkpoints it will be possible to prune the search query/indices and e.g. avoid calling out to frozen/cold indices.

### PR DESCRIPTION
**Pull request by hendrikmuhs**
Use point in time API for every checkpoint in transform. Using point in time reduces pressure on the source indexes, e.g. less refreshes. In case, pit isn't supported (e.g. when searching remote clusters) it falls back to ordinary search requests as before.

closes #73481

### CODE REVIEW

**Comment by hendrikmuhs** [comment]
retest this please

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java [hunk 1]
```diff
@@ -440,6 +450,132 @@ SeqNoPrimaryTermAndIndex getSeqNoPrimaryTermAndIndex() {
...
+    }
+
+    @Override
+    protected void onStop() {
+        closePointInTime();
+        super.onStop();
+    }
+
+    private void closePointInTime() {
+        if (pit != null) {
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:466]: [nit] We could save a little bit on indentation here:
```
if (pit == null) {
   return;
}
...
```

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java [hunk 2]
```diff
@@ -440,6 +450,132 @@ SeqNoPrimaryTermAndIndex getSeqNoPrimaryTermAndIndex() {
...
+            String oldPit = pit.getEncodedId();
+            pit = null;
+            ClosePointInTimeRequest closePitRequest = new ClosePointInTimeRequest(oldPit);
+            ClientHelper.executeWithHeadersAsync(
+                transformConfig.getHeaders(),
+                ClientHelper.TRANSFORM_ORIGIN,
+                client,
+                ClosePointInTimeAction.INSTANCE,
+                closePitRequest,
+                ActionListener.wrap(response -> { logger.trace("[{}] closed pit search context [{}]", getJobId(), oldPit); }, e -> {
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:476]:
```suggestion
                ActionListener.wrap(response -> {
                    logger.trace("[{}] closed pit search context [{}]", getJobId(), oldPit);
                },
                e -> {
```

**Review by przemekwitek — APPROVED** [review]
LGTM

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java [hunk 3]
```diff
@@ -111,13 +120,14 @@ protected void doNextSearch(long waitTimeInNanos, ActionListener<SearchResponse>
...
         }
-        ClientHelper.executeWithHeadersAsync(
-            transformConfig.getHeaders(),
-            ClientHelper.TRANSFORM_ORIGIN,
-            client,
-            SearchAction.INSTANCE,
+
+        if (getNextCheckpoint().getCheckpoint() != pitCheckpoint) {
+            closePointInTime();
+        }
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:126]: It seems to me that we should not move this execution thread forward until the `pit` is closed.

It is conceivable right (though unlikely) that this `closePointInTime()` is executing, but `doSearch` is being handled and consequently, we close the wrong PIT and leave one left over.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java [hunk 4]
```diff
@@ -440,6 +450,134 @@ SeqNoPrimaryTermAndIndex getSeqNoPrimaryTermAndIndex() {
...
+        ClientHelper.executeWithHeadersAsync(
+            transformConfig.getHeaders(),
+            ClientHelper.TRANSFORM_ORIGIN,
+            client,
+            ClosePointInTimeAction.INSTANCE,
+            closePitRequest,
+            ActionListener.wrap(response -> { logger.trace("[{}] closed pit search context [{}]", getJobId(), oldPit); }, e -> {
+                // note: closing the pit should never throw, even if the pit is invalid
+                logger.error(new ParameterizedMessage("[{}] Failed to close point in time reader", getJobId()), e);
+            })
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:482]: I think the `logger.trace` should have a message supplier like `() -> new ParameterizedMessage` to prevent strings from being created when `trace` is disabled.

Not a huge deal as this is not a "hot path"

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java [hunk 5]
```diff
@@ -440,6 +450,134 @@ SeqNoPrimaryTermAndIndex getSeqNoPrimaryTermAndIndex() {
...
+            transformConfig.getHeaders(),
+            ClientHelper.TRANSFORM_ORIGIN,
+            client,
+            OpenPointInTimeAction.INSTANCE,
+            pitRequest,
+            ActionListener.wrap(response -> {
+                pit = new PointInTimeBuilder(response.getPointInTimeId()).setKeepAlive(PIT_KEEP_ALIVE);
+                searchRequest.source().pointInTimeBuilder(pit);
+                pitCheckpoint = getNextCheckpoint().getCheckpoint();
+                logger.trace("[{}] using pit search context with id [{}]", getJobId(), pit.getEncodedId());
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:511]: Similar comment `() -> new ParameterizedMessage` seems better to me for `trace`

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java [hunk 6]
```diff
@@ -57,21 +64,24 @@
...
 class ClientTransformIndexer extends TransformIndexer {
 
+    private static final TimeValue PIT_KEEP_ALIVE = TimeValue.timeValueSeconds(30);
     private static final Logger logger = LogManager.getLogger(ClientTransformIndexer.class);
 
     private final Client client;
     private final AtomicBoolean oldStatsCleanedUp = new AtomicBoolean(false);
 
     private final AtomicReference<SeqNoPrimaryTermAndIndex> seqNoPrimaryTermAndIndex;
+    private PointInTimeBuilder pit;
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:74]: does `pit` and `disablePit` need to be `volatile`? They are accessed from separate threads in different execution paths.

Comment (hendrikmuhs) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:74]: pit is always accessed by the indexer thread, even the `onStop` call originates from the indexer, not from the `_stop` transport if that's what you mean

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java (hunk 3 above)
Comment (hendrikmuhs) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:126]: pit is copied and set to `null` in the sync part of `closePointInTime()`, see line 470++. So you are right that we might open a new pit while still closing the other, however that's allowed and I don't see a race condition that could lead to mixing up the two.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java (hunk 4 above)
Comment (hendrikmuhs) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:482]: I don't think this applies. This is only a problem if one or more arguments needs to be constructed, e.g. if `getJobId()` would build the id and therefore execute something. This is not the case.

The message string itself gets only constructed _after_ the check whether trace is enabled or not.

I wish we have static code analysis for this this, it's such a common problem.

**Comment by hendrikmuhs** [comment]
@elasticmachine update branch

**Review by benwtrent — APPROVED** [review]

**Comment by hendrikmuhs** [comment]
## Test results

Using `pit` refreshes on the source index can be reduced significantly:

image

This chart compares a baseline (indexing without doing any searches/transforms), transform without pit (`<7.15`) and transform using pit (`>7.15`).

The usage of `pit` reduces the number of refreshes significantly, correlated to that the number of merges goes down from `1440` to `560` and the time spend merging from `17.1` minutes to `9.1` minutes.

**It depends:** This is not a representative benchmark. The benefit of using `pit` depends on data, ingest rates, the transform configuration and other query executors like dashboards that use the same source index.

In summary you might not see a resource reduction in the same order of magnitude with your data, however pit should reduce overhead for all use cases where the source index isn't static (continuous transform).
